freqtrade/rpc/external_signal/controller.py

```diff
@@ -0,0 +1,449 @@
+        """
+        type = message.get("data_type", LeaderMessageType.default)
+        data = message.get("data")
+
+        handler: Callable = self._message_handlers[type]
+        handler(type, data)
+
+    # ----------------------------------------------------------------------
+
+    async def run_leader_mode(self):
+        """
+        Main leader coroutine
+
+        This starts all of the leader coros and registers the endpoint on
+        the ApiServer
+        """
+        self.register_leader_endpoint()
+        self.log_api_token()
+
+        self._sub_tasks = [
+            self._loop.create_task(self._broadcast_queue_data())
+        ]
+
+        return await asyncio.gather(*self._sub_tasks)
+
+    async def run_follower_mode(self):
+        """
+        Main follower coroutine
+
+        This starts all of the follower connection coros
+        """
+
+        rpc_lock = asyncio.Lock()
+
+        self._sub_tasks = [
+            self._loop.create_task(self._handle_leader_connection(leader, rpc_lock))
+            for leader in self.leaders_list
+        ]
+
+        return await asyncio.gather(*self._sub_tasks)
+
+    async def _broadcast_queue_data(self):
+        """
+        Loop over queue data and broadcast it
+        """
+        # Instantiate the queue in this coroutine so it's attached to our loop
+        self._queue = ThreadedQueue()
+        async_queue = self._queue.async_q
+
+        try:
+            while self._running:
+                # Get data from queue
+                data = await async_queue.get()
+
+                # Broadcast it to everyone
+                await self.channel_manager.broadcast(data)
+
+                # Sleep
+                await asyncio.sleep(self.push_throttle_secs)
+
+        except asyncio.CancelledError:
+            # Silently stop
+            pass
+
+    async def get_api_token(
+        self,
+        websocket: FastAPIWebSocket,
+        token: Union[str, None] = None
+    ):
+        """
+        Extract the API key from query param. Must match the
+        set secret_api_key or the websocket connection will be closed.
+        """
+        if token == self.secret_api_key:
+            return token
+        else:
+            logger.info("Denying websocket request...")
+            await websocket.close(code=status.WS_1008_POLICY_VIOLATION)
+
+    def register_leader_endpoint(self, path: str = "/signals/ws"):
+        """
+        Attach and start the main leader loop to the ApiServer
+
+        :param path: The endpoint path
+        """
+        if not self.api_server:
+            raise RuntimeError("The leader needs the ApiServer to be active")
+
+        # The endpoint function for running the main leader loop
+        @self.api_server.app.websocket(path)
+        async def leader_endpoint(
+            websocket: FastAPIWebSocket,
+            api_key: str = Depends(self.get_api_token)
+        ):
+            await self.leader_endpoint_loop(websocket)
```

Comment:
The websocket should be registered as part of a default webserver (so the websocket endpoint is always there the moment you start the webserver).

Websockets have other usages too - but this architecture limits the usage of this too much.

That'll obviously also require a method to authenticate the websocket - most likely by checking for either the tokens generated here - or for the JWT tokens used in the frontend.

Reply:
One main thing that the Leader needs as well as the endpoint is the coro `_broadcast_queue_data` to be running in the background. So any data added to the queue gets automatically pushed through to the `ChannelManager.broadcast`.

Reply:
that's not necessarily a problem, is it? 
That same structure will also be needed for other messagetypes (think about sending "entry" or "exit" messages via that route).

We still just call `.send_msg()` with appropriate message types - the rpc handler (apiserver, in this case) places the stuff in the queue, which is then processed technically immediately and sent to all subscribed clients.

Now i'm not sure if that's done already - but you'll want the client to actively subscribe to certain topics / message types (obviously the client here would subscribe to this messagetype).

A webUI may not need all analyzed data force-fed to it - but may want to only listen to buy/sell notifications.